clang/lib/Basic/Targets/SPIR.cpp

```diff
@@ -35,8 +35,10 @@ static constexpr Builtin::Info BuiltinInfos[] = {
 static_assert(std::size(BuiltinInfos) == NumBuiltins);
 
 llvm::SmallVector<Builtin::InfosShard>
-SPIRVTargetInfo::getTargetBuiltins() const {
-  return {{&BuiltinStrings, BuiltinInfos}};
+BaseSPIRTargetInfo::getTargetBuiltins() const {
+  if (getTriple().isSPIRV())
+    return {{&BuiltinStrings, BuiltinInfos}};
```

Comment:
I don't understand why this change (in SPIR.cpp and SPIR.h) is necessesary.  The existing code accomplished the same thing with method overriding. The original code also had some defensive programming safety to me. We went  from a compile time enforcement of  SPIRVTargetInfo== BuiltinsSPIRV.td to now a runtime enforcement. The problem with that is the next guy that comes along and makes a change might assume the builtins so far are SPIRV and thats why you are checking the triple, then add a new SPIR flavor builtin that would not be valid for SPIRV.

Reply:
Oh, that's a left over of my initial prototype in intel/llvm sorry. For historical reasons `spir` / `spir64` is also used there, so had to hack a bit to allow proper testing there. The change should be moving it to `BaseSPIRVTargetInfo` to only enable `spirv`, `spirv32` and `spirv64`. I updated the PR.

Reply:
Ok we might have a hierarchy problem then how do we say a builtin is valid for `spirv` but not `spirv32` or `spriv64` and vice versa. see https://github.com/llvm/llvm-project/pull/137805#discussion_r2066520947

Reply:
I had a few thought about this, but there is a bit more to it:

1. Vulkan vs OpenCL environment (and for which target version)
2. Enabled vs disabled extensions
3. SPIR-V version

each will have an impact on the set of enabled builtins. A split between common, Vk env (`spirv`) and OpenCL env (`spirv32` / `spirv64`) builtins is probably a good step towards answering (1) by construction. What I don't like though is you'll end up with an "unknown builtin" rather than a clear message.
For the rest, I wanted to create an RFC to suggest the use of target features as there is plenty of existing infrastructure for this job.

Reply:
What if we had a `BuiltinsSPIRVCore.td` `BuiltinsSPIRVVK.td` and a `BuiltinsSPIRVCL.td` The stuff shared between CL and VK could live in Core and the VK and CL builtins will include the Core one. That should let us have seperate `SPIRVTargetInfo` and `SPIRV64TargetInfo`/`SPIRV32TargetInfo`. You won't have an unknown builtin if we do it this way because we would never expose the builtins that aren't valid for the target.

Reply:
I'll consider this issue resolved though if you plan to address it via an RFC.

Reply:
> What if we had a BuiltinsSPIRVCore.td BuiltinsSPIRVVK.td and a BuiltinsSPIRVCL.td

I was looking into doing this.

> You won't have an unknown builtin if we do it this way because we would never expose the builtins that aren't valid for the target.

You will if you use a VK one when targeting `spirv64` rather than a clear message. But it is nitpicking and I only have a weak preference on the matter.

Reply:
@farzonl Apply the suggestion, although you can't really split into 2 distinct set cleanly (common + vk and common + cl) due to the ID assignment (part is done dynamically). So I went for a unique set and diagnose when vk builtins are used with spirv32 or spirv32 or when cl builtins are used with spirv.